api/src/main/java/org/openmrs/api/impl/DomainServiceImpl.java

```diff
@@ -0,0 +1,185 @@
+     */
+    @Autowired
+    public DomainServiceImpl(List<OpenmrsService> services) {
+        if (services != null) {
+            for (OpenmrsService service : services) {
+                if (this.getClass().isAssignableFrom(service.getClass())) {
+                    continue;
+                }
+                for (Method method : service.getClass().getMethods()) {
+                    if (method.getName().startsWith("get") &&
+                        method.getName().endsWith("ByUuid") &&
+                        method.getParameterCount() == 1 &&
+                        method.getParameterTypes()[0].equals(String.class)) {
+
+                        Class<?> domainKey = method.getReturnType();
+                        if (!domainFetchers.containsKey(domainKey)) {
+                            domainFetchers.put(domainKey, new DomainFetcher(method.getReturnType(), service, method));
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Retrieves an object of the given type using its UUID.
+     *
+     * @param type the expected type of the domain object
+     * @param uuid the UUID of the domain object to fetch
+     * @param <T>  a type that matches the domain object class
+     * @return an instance of the requested type
+     * @throws RuntimeException if no fetcher exists for the given type
+     * @throws ClassCastException if the fetched object is not of the expected type
+     */
+    @Transactional(readOnly = true)
+    public <T> T fetchByUuid(Class<T> type, String uuid) {
+        Object result = null;
+
+        DomainFetcher fetcher = domainFetchers.get(type);
+        if (fetcher != null) {
+            result = fetcher.fetch(uuid);
+        } else {
+            throw new RuntimeException("No suitable fetcher found for domain: " + type);
+        }
+
+        if (!type.isInstance(result)) {
+            throw new ClassCastException("Expected: " + type + ", but got: " + (result != null ? result.getClass() : "null"));
+        }
+        return type.cast(result);
+    }
+
+    /**
+     * Returns a list of all domain types that are currently registered and
+     * resolvable by this service.
+     *
+     * @return a list of registered domain classes
+     */
+    public List<Class<?>> getDomainTypes() {
+        Set<Class<?>> types = new HashSet<>();
+        for (DomainFetcher fetcher : domainFetchers.values()) {
+            types.add(fetcher.getReturnType());
+        }
+        return new ArrayList<>(types);
+    }
+
+    /**
+     * Internal helper that holds metadata and invocation logic for a specific
+     * {@code getXByUuid(String)} method.
+     */
+    private class DomainFetcher {
+
+        private final Class<?> returnType;
+        private final OpenmrsService service;
+        private final Method method;
+
+        public DomainFetcher(Class<?> returnType, OpenmrsService service, Method method) {
```

Comment:
I don't like using reflection due to performance reasons. It may hit us with files using hundreds of references. No exact figures for service calls, but simple tests show it's a few times slower than usual method calls. There's much faster LambdaMetafactory that we could use if we want to go that route. However, I would be in favour of introducing an interface to our services:
```
public interface RefByUuid {
	
	<T> T getRefByUuid(Class<T> type, String uuid);
	
	List<?> getRefTypes();
}
```
and **maybe** use reflection only if the interface is missing, but if I were to decide, I would not use it.